File: src/animate/color.ts
import { Canvas } from 'meo-skia-canvas'

/**
 * A colour resolved to sRGB. Channels are 0–255 and alpha is 0–1.
 *
 * Channels are floats and may fall outside 0–255: that is how a colour sRGB cannot reach is
 * carried. Display P3's red is `r` above 255 with `g` and `b` slightly negative — extended sRGB,
 * the same colour named in sRGB's coordinates. Clamping here would quietly substitute the duller
 * sRGB red for the one that was asked for.
 */
export interface Rgba {
  r: number
  g: number
  b: number
  a: number
}

const CHANNEL_MAX = 255

/**
 * Two sentinels used to tell a rejected colour from an accepted one.
 *
 * Assigning an unparseable value to `fillStyle` leaves the previous value in place instead of
 * raising — the behaviour the HTML canvas spec requires. A single probe therefore cannot tell
 * "invalid" from "valid and equal to whatever was already set". Probing twice from two different
 * starting points can: a colour the engine accepts produces the same answer from both, while a
 * rejected one leaves each sentinel untouched and the two answers differ.
 */
const PROBE_A = '#000000'
const PROBE_B = '#ffffff'

/**
 * Scratch canvas used to make the engine do the parsing.
 *
 * Created once and kept: it is a single pixel, and building one per call would dominate the cost
 * of interpolating a colour sixty times a second.
 */
let scratch: Canvas | null = null

function surface(): Canvas {
  if (!scratch) {
    // Float, not the default eight bits a channel. An eight-bit surface clamps on the way in, so a
    // Display P3 red would land on plain sRGB red before anything could read it back — the two
    // become the same colour, and the gamut is gone before the parse finishes.
    scratch = new Canvas(1, 1, { colorType: 'RGBAF32' })
  }
  return scratch
}

function context(): ReturnType<Canvas['getContext']> {
  return surface().getContext('2d')
}

/**
 * How many parsed colours are kept.
 *
 * Bounded because the cache outlives every render: it is module state, so without a limit a
 * long-running server would accumulate one permanent entry per distinct colour it ever drew. A
 * render that computes a colour per page — which is what a chained `mix()` produces — makes that a
 * steady leak rather than a theoretical one. A few thousand entries is far more than any scene's
 * palette and costs a few hundred kilobytes.
 */
export const COLOR_CACHE_LIMIT = 4096

/**
 * Parsed colours, keyed by the exact input string, most recently used last.
 *
 * An animation asks for the same two endpoints on every page, so this turns a per-page parse into
 * a map lookup. Entries are frozen, and callers get a copy — a mutable cached object would let one
 * caller corrupt every later reader.
 *
 * Eviction is least-recently-used rather than oldest-first, because age says nothing about whether
 * a colour is still in play: an animation's endpoints are the oldest entries it has and also the
 * two it reads on every single page.
 */
const cache = new Map<string, Readonly<Rgba>>()

/** Number of colours currently cached. Exposed for tests, which is why it is a function. */
export function colorCacheSize(): number {
  return cache.size
}

/**
 * Resolves any colour the rendering engine accepts into sRGB.
 *
 * The engine does the work: the colour is painted onto a one-pixel canvas and read back. That is
 * deliberately not a CSS parser — it means every syntax the engine supports is supported here for
 * free, today and after any upgrade, including `lab()`, `oklch()` and `color(display-p3 …)`, which
 * a hand-written parser would either miss or approximate.
 *
 * Wide-gamut inputs survive: the scratch surface holds floats, so a colour beyond sRGB comes back
 * as extended sRGB rather than clipped to the edge of it.
 * @throws if the engine does not recognise the colour, rather than returning a misleading black.
 */
export function parseColor(css: string): Rgba {
  const hit = cache.get(css)
  if (hit) {
    // Re-inserted so it moves to the end: a Map iterates in insertion order, which is what makes
    // "the first key" the least recently used one.
    cache.delete(css)
    cache.set(css, hit)
    return { ...hit }
  }

  const ctx = context()

  ctx.fillStyle = PROBE_A
  ctx.fillStyle = css
  const fromA = String(ctx.fillStyle)

  ctx.fillStyle = PROBE_B
  ctx.fillStyle = css
  const fromB = String(ctx.fillStyle)

  if (fromA !== fromB) {
    throw new Error(`[canvas] "${css}" is not a colour the renderer recognises`)
  }

  const parsed = Object.freeze(fromNormalized(fromA, ctx))

  if (cache.size >= COLOR_CACHE_LIMIT) {
    const oldest = cache.keys().next().value
    if (oldest !== undefined) cache.delete(oldest)
  }
  cache.set(css, parsed)

  return { ...parsed }
}

/** `#rrggbb` — what the engine returns for any opaque colour already inside sRGB. */
const HEX_RE = /^#([0-9a-f]{6})$/i
/** `rgba(r, g, b, a)` — what it returns once alpha is involved. */
const RGBA_RE = /^rgba?\(\s*([\d.]+)\s*,\s*([\d.]+)\s*,\s*([\d.]+)\s*(?:,\s*([\d.]+)\s*)?\)$/i
/** `color(space c1 c2 c3 [/ a])` — what it returns for wide-gamut and out-of-gamut colours. */
const COLOR_FN_RE = /^color\((\S+)\s+([\d.eE+-]+)\s+([\d.eE+-]+)\s+([\d.eE+-]+)\s*(?:\/\s*([\d.eE+-]+)\s*)?\)$/

/**
 * Turns the engine's normalised colour string into sRGB channels.
 *
 * Read from the string wherever possible rather than from a painted pixel, because painting is
 * lossy in one specific and common case: the canvas stores premultiplied alpha, so a fully
 * transparent colour comes back as transparent black and loses its hue. Fading in from
 * `rgba(255, 0, 0, 0)` would then start from black instead of red. The string keeps the channels.
 *
 * Only a `color()` result needs the engine's help, since converting a space like `display-p3` into
 * sRGB is exactly the arithmetic worth not reimplementing. Even then alpha comes from the string,
 * and the colour is painted opaque so premultiplication cannot touch the channels.
 */
function fromNormalized(normalized: string, ctx: ReturnType<Canvas['getContext']>): Rgba {
  const hex = HEX_RE.exec(normalized)
  if (hex) {
    const value = parseInt(hex[1], 16)
    return { r: (value >> 16) & 0xff, g: (value >> 8) & 0xff, b: value & 0xff, a: 1 }
  }

  const rgba = RGBA_RE.exec(normalized)
  if (rgba) {
    return {
      r: Number(rgba[1]),
      g: Number(rgba[2]),
      b: Number(rgba[3]),
      a: rgba[4] === undefined ? 1 : Number(rgba[4]),
    }
  }

  const fn = COLOR_FN_RE.exec(normalized)
  if (fn) {
    const [, space, c1, c2, c3, alpha] = fn

    // The only form that needs the surface. `#rrggbb` and `rgba()` are already sRGB and already
    // exact in the string; reading those back through float32 would return 102.0000015 for 102 and
    // buy nothing. A `color()` result is the one that may name a colour sRGB cannot reach, and
    // converting a space like display-p3 is the arithmetic worth not reimplementing.
    //
    // Painted opaque so premultiplied storage cannot touch the channels: at alpha 0 the pixel would
    // otherwise come back transparent black with the hue gone. Alpha comes from the string, which
    // has it regardless.
    ctx.clearRect(0, 0, 1, 1)
    ctx.fillStyle = `color(${space} ${c1} ${c2} ${c3})`
    ctx.fillRect(0, 0, 1, 1)

    const raw = surface().toBufferSync('raw', { colorType: 'RGBAF32' })

    return {
      r: raw.readFloatLE(0) * CHANNEL_MAX,
      g: raw.readFloatLE(4) * CHANNEL_MAX,
      b: raw.readFloatLE(8) * CHANNEL_MAX,
      a: alpha === undefined ? 1 : Number(alpha),
    }
  }

  throw new Error(`[canvas] the renderer returned a colour this does not understand: "${normalized}"`)
}

/** Whether the engine recognises this colour. Never throws. */
export function isColor(css: string): boolean {
  try {
    parseColor(css)
    return true
  } catch {
    return false
  }
}

const clampChannel = (value: number): number => Math.min(CHANNEL_MAX, Math.max(0, Math.round(value)))

/** Whether every channel sits inside what sRGB can express, and so inside what hex can write. */
const inGamut = ({ r, g, b }: Rgba): boolean => [r, g, b].every(c => c >= 0 && c <= CHANNEL_MAX)

/** Digits kept for an out-of-gamut channel: enough that a float32 round-trips visually intact. */
const EXTENDED_PRECISION = 6

/**
 * Writes a colour back as a string the engine accepts.
 *
 * Ordinary colours become hex, or `rgba()` once alpha is involved, because that is how this
 * library's props are written everywhere else and it stays readable.
 *
 * A colour outside sRGB cannot be written either way — hex has no room for a channel above 255 or
 * below 0 — so it becomes `color(srgb …)`, which carries the extended values verbatim and is read
 * back exactly. Clamping instead would substitute a duller colour without saying so.
 */
export function formatColor(color: Rgba): string {
  const alpha = Math.min(1, Math.max(0, color.a))

  if (!inGamut(color)) {
    const [red, green, blue] = [color.r, color.g, color.b].map(c => Number((c / CHANNEL_MAX).toFixed(EXTENDED_PRECISION)))
    const tail = alpha >= 1 ? '' : ` / ${Number(alpha.toFixed(EXTENDED_PRECISION))}`
    return `color(srgb ${red} ${green} ${blue}${tail})`
  }

  const [red, green, blue] = [color.r, color.g, color.b].map(clampChannel)

  if (alpha >= 1) {
    return `#${[red, green, blue].map(c => c.toString(16).padStart(2, '0')).join('')}`
  }
  // Trimmed so a half-way alpha reads as 0.5 rather than 0.5000000000000001.
  return `rgba(${red}, ${green}, ${blue}, ${Number(alpha.toFixed(3))})`
}

/**
 * Blends two colours, in any accepted format, and returns one string.
 *
 * Interpolation happens in sRGB with a straight alpha, which is what the canvas compositing model
 * expects. `t` is clamped, so a track that overshoots cannot produce an impossible colour.
 */
export function mixColor(from: string, to: string, t: number): string {
  const clamped = Math.min(1, Math.max(0, t))
  const a = parseColor(from)
  const b = parseColor(to)

  return formatColor({
    r: a.r + (b.r - a.r) * clamped,
    g: a.g + (b.g - a.g) * clamped,
    b: a.b + (b.b - a.b) * clamped,
    a: a.a + (b.a - a.a) * clamped,
  })
}

File: crates/meo-canvas-core/src/animate/color.rs
//! Colour as an animatable value.
//!
//! `f64` channels, unclamped, so an interpolation can sit between two colours
//! and an overshoot outside both; narrowing to bytes happens once, where the
//! value becomes a scene colour. `r`, `g` and `b` run 0 to 255 and `a` 0 to 1,
//! the scale both surfaces use and the one `in_gamut` needs to send an
//! out-of-range colour to `color(srgb ...)`. [`mix`] does not clamp `t`: CSS
//! interpolates through an overshooting timing function and clamps at paint,
//! which `outBack` and `outElastic` rely on.

#![expect(
    clippy::suboptimal_flops,
    reason = "compared bit-for-bit against v9's own numbers; see \
              `animate::easing` for the rule and where it does not apply."
)]

use meo_canvas_scene::style::paint::Color;

/// The top of a colour channel: 255, the scale both surfaces use.
pub const CHANNEL_MAX: f64 = 255.0;

/// A colour with room to be between two others, and outside both.
#[derive(Debug, Clone, Copy, PartialEq, Default)]
pub struct Rgba {
    /// Red, 0 to 255, and not clamped to it.
    pub r: f64,
    /// Green, likewise.
    pub g: f64,
    /// Blue, likewise.
    pub b: f64,
    /// Alpha, 0 to 1, and not clamped to it either.
    pub a: f64,
}

impl Rgba {
    /// Whether every channel sits inside what sRGB can express.
    ///
    /// Alpha is not part of it: alpha outside 0..1 is meaningless rather than
    /// out of gamut, and is clamped wherever it is used.
    #[must_use]
    pub fn in_gamut(self) -> bool {
        [self.r, self.g, self.b]
            .iter()
            .all(|channel| (0.0..=CHANNEL_MAX).contains(channel))
    }

    /// The scene colour this becomes when drawn: the only clamp, since a
    /// channel outside the gamut has nowhere to go in a byte.
    #[must_use]
    pub fn to_color(self) -> Color {
        let byte =
            |channel: f64| (channel.clamp(0.0, 1.0) * 255.0).round() as u8;
        Color::rgba(byte(self.r), byte(self.g), byte(self.b), byte(self.a))
    }
}

/// Blends two colours, straight alpha, in sRGB. `t` is not clamped, so an
/// overshooting curve carries the colour out of the gamut until paint.
#[must_use]
pub fn mix(from: Rgba, to: Rgba, t: f64) -> Rgba {
    let between = |a: f64, b: f64| a + (b - a) * t;
    Rgba {
        r: between(from.r, to.r),
        g: between(from.g, to.g),
        b: between(from.b, to.b),
        a: between(from.a, to.a),
    }
}

#[cfg(test)]
mod tests {
    use super::{Rgba, mix};

    /// Mid grey, opaque, as channels rather than bytes.
    const GREY: Rgba = Rgba {
        r: 0.5,
        g: 0.5,
        b: 0.5,
        a: 1.0,
    };

    #[test]
    fn a_mix_past_the_end_leaves_the_gamut_rather_than_stopping() {
        // The whole reason this type is not four bytes: an overshooting curve
        // hands `t` past 1, and `mix` carries it out of the gamut.
        let white = Rgba {
            r: 255.0,
            g: 255.0,
            b: 255.0,
            a: 1.0,
        };
        let over = mix(GREY, white, 1.5);
        assert!(over.r > 255.0, "the mix stopped at the endpoint");
        assert!(!over.in_gamut());
        // And it comes back at the byte boundary rather than in between.
        assert_eq!(over.to_color().r, 255);
    }

    #[test]
    fn a_mix_inside_the_range_is_ordinary() {
        let black = Rgba {
            r: 0.0,
            g: 0.0,
            b: 0.0,
            a: 1.0,
        };
        let white = Rgba {
            r: 1.0,
            g: 1.0,
            b: 1.0,
            a: 1.0,
        };
        let half = mix(black, white, 0.5);
        assert_eq!(half, GREY);
        assert!(half.in_gamut());
        assert_eq!(half.to_color().r, 128);
    }

    #[test]
    fn alpha_is_not_part_of_the_gamut_but_is_part_of_the_clamp() {
        // Alpha outside 0..1 is meaningless rather than out of gamut: the
        // colour is still sRGB, so `in_gamut` ignores it and `to_color`
        // clamps it like any other channel.
        let loud = Rgba { a: 1.4, ..GREY };
        assert!(loud.in_gamut());
        assert_eq!(loud.to_color().a, 255);
        let quiet = Rgba { a: -0.2, ..GREY };
        assert!(quiet.in_gamut());
        assert_eq!(quiet.to_color().a, 0);
    }

    #[test]
    fn a_negative_channel_is_out_of_gamut_at_both_ends() {
        assert!(!Rgba { r: -0.01, ..GREY }.in_gamut());
        assert!(!Rgba { b: 255.01, ..GREY }.in_gamut());
        assert_eq!(Rgba { r: -20.0, ..GREY }.to_color().r, 0);
    }
}
